src/pybind/mgr/volumes/fs/clone_stats.py

```diff
@@ -0,0 +1,215 @@
+
+    def __init__(self, volname):
+        self.volname = volname
+
+        self.src_group_name = None
+        self.src_subvol_name = None
+        self.src_path = None
+
+        self.dst_group_name = None
+        self.dst_subvol_name = None
+        self.dst_path = None
+
+
+class CloneProgressReporter(AsyncJobProgressReporter):
+    '''
+    Report progress made by cloner threads.
+    '''
+
+    def __init__(self, volclient, volspec):
+        super().__init__(volclient, volspec)
+
+        # need to figure out how many progress bars should be printed. print 1
+        # progress bar if number of ongoing clones is less than this value,
+        # else print 2.
+        self.max_concurrent_clones = self.volclient.mgr.max_concurrent_clones
+
+        self.op_name = 'clone'
+
+    def _get_clone_dst_info(self, fs_handle, ci, clone_entry,
+                            clone_index_path):
+        log.debug('collecting info for cloning destination')
+
+        ce_path = os_path_join(clone_index_path, clone_entry)
+        # XXX: This may raise ObjectNotFound exception. As soon as cloning is
+        # finished, clone entry is deleted by cloner thread. This exception is
+        # handled in _get_info_for_all_clones().
+        dst_subvol_base_path = fs_handle.readlink(ce_path, PATH_MAX).\
+            decode('utf-8')
+
+        ci.dst_group_name, ci.dst_subvol_name = \
+            resolve_group_and_subvolume_name(self.volspec, dst_subvol_base_path)
+        with open_subvol_in_vol(self.volclient, self.volspec, ci.volname,
+                                ci.dst_group_name, ci.dst_subvol_name,
+                                SubvolumeOpType.CLONE_INTERNAL) \
+                                as (_, _, dst_subvol):
+            ci.dst_path = dst_subvol.path
+            log.debug(f'destination subvolume path for clone - {ci.dst_path}')
+
+        log.debug('finished collecting info for cloning destination')
+
+    def _get_clone_src_info(self, fs_handle, ci):
+        log.debug('collecting info for cloning source')
+
+        with open_clone_subvol_pair_in_vol(self.volclient, self.volspec,
+                                           ci.volname, ci.dst_group_name,
+                                           ci.dst_subvol_name) as \
+                                           (dst_subvol, src_subvol, snap_name):
+            ci.src_group_name = src_subvol.group_name
+            ci.src_subvol_name = src_subvol.subvolname
+            ci.src_path = src_subvol.snapshot_data_path(snap_name)
+            log.debug(f'source subvolume path for clone - {ci.src_path}')
+
+        log.debug('finished collecting info for cloning source')
+
+    def _get_info_for_all_clones(self):
+        clones:list[CloneInfo] = []
+
+        log.debug('collecting all entries in clone index...')
+        volnames = list_volumes(self.volclient.mgr)
+        for volname in volnames:
+            with open_volume_lockless(self.volclient, volname) as fs_handle:
```

Comment:
Forgot to post this, this code wasn't added in this PR but moved from `stats_util.py` to `clone_stats.py`. Therefore I don't intend to make improvements on this PR since focus here is add to new feature. Marking this as resolved.